### ISSUE TITLE
TC003 and TC008 false positives for TypeAlias

### ISSUE THREAD

**Issue by sporksmith** [snok/flake8-type-checking#209]
Given the following content of `test.py`:

```
from __future__ import annotations

from collections.abc import Mapping
from typing import TYPE_CHECKING, Union, TypeAlias

MyTypeAlias: TypeAlias = Union[
    bool,
    str,
    int,
    float,
    Mapping[str, "MyTypeAlias"],
]
```

I get the following flake8 TC failures:

```
test.py:3:1: TC003 Move built-in import 'collections.abc.Mapping' into a type-checking block
test.py:11:18: TC008 Type alias 'MyTypeAlias' does not need to be a string literal
```

If I "fix" either of these, executing `test.py` fails at runtime (on python 3.10)

**Comment by Daverball** [comment]
Alright, the `TC008` is a bug with recursive type aliases, which should be fixed via the linked PR.

The `TC003` violation however is not a bug, the idea is to encourage people to partially stringize their `TypeAlias` or move them into `if TYPE_CHECKING` blocks, if it lets them move some of the imports inside a type checking block. This is a little opinionated, so I would be open to a feature request which adds a setting to change this behavior (i.e. a setting for which type aliases will always be considered runtime uses, so you will never get a TC001-003 for any symbols used within them).

So while fixing the `TC003` will result in broken code, you should be getting a `TC007` violation on this new code, which when fixed will lead to working code.

**Comment by Daverball** [comment]
flake8-type-checking v3.2.0 was just released which contains this bugfix

**Comment by sporksmith** [comment]
> Alright, the TC008 is a bug with recursive type aliases, which should be fixed via the linked PR.

Awesome, thanks!

> The TC003 violation however is not a bug, the idea is to encourage people to partially stringize their TypeAlias or move them into if TYPE_CHECKING blocks

I'm not sure what you mean by partially stringizing the TypeAlias.

Moving the whole alias definition into a TYPE_CHECKING block is undesirable in my case, since I don't want to force other modules importing the TypeAlias to themselves do so inside a TYPE_CHECKING block. I imagine it'd be even more undesirable if it were something meant to be exposed for use in other packages.

For more context, I just started using TYPE_CHECKING blocks in my project, and am currently leaning towards doing it only as much as I need to to break import cycles. I've actually disabled `TC002` and `TC003` altogether for my project. My motivation for using flake8-type-checking is mostly to use `TC009` to statically detect if guarded imports are used outside of type-checking (and would therefore result in a runtime error).  https://gitlab.torproject.org/tpo/core/chutney/-/merge_requests/116

So the remaining `TC003` violation doesn't really affect me personally at the moment, but my humble opinion is that it shouldn't be a violation :)

Thanks again!

**Comment by Daverball** [comment]
> I'm not sure what you mean by partially stringizing the TypeAlias.

For the example you have given this would mean changing it to the following:
```python
from __future__ import annotations

from typing import TYPE_CHECKING, Union, TypeAlias
if TYPE_CHECKING:
    from collections.abc import Mapping

MyTypeAlias: TypeAlias = Union[
    bool,
    str,
    int,
    float,
    "Mapping[str, MyTypeAlias]",
]
```

The type alias would still be available at runtime. So this would only be problematic if you use some kind of runtime introspection on the type alias and need all of its value expression components to be available at runtime, although recursive type aliases usually don't work for that use-case anyways unless the runtime introspection library specifically makes it work.

But overall this behavior is definitely opinionated, so it's completely fine to disable `TC001-003`, if you don't care about getting rid of  import statements that are only there for type hints.

FWIW the implementation of these rules in Ruff is less opinionated (by design, since there are auto fixes in Ruff, so there is less of an opportunity to make the call on whether or not the violation is actually valid and adding a noqa comment instead), so `TC001-003` will not trigger for type aliases, it will also by default never make you put an annotation in quotes, just so you can move it into a type checking block. So that implementation might be more to your liking. On the flip-side there's no rule beside UP037, that will help you get rid of redundant quotes in type hints. If you're using `from __future___ import annotations` however, this is less of a problem, since most of the redundant quotes will still be removed.

**Comment by sporksmith** [comment]
> For the example you have given this would mean changing it to the following:

Ah, that makes sense. Thanks!
